src/core/lib/event_engine/posix_engine/posix_endpoint.cc

```diff
@@ -0,0 +1,1262 @@
+#include <netinet/in.h>  // IWYU pragma: keep
+
+#ifndef SOL_TCP
+#define SOL_TCP IPPROTO_TCP
+#endif
+
+#ifndef TCP_INQ
+#define TCP_INQ 36
+#define TCP_CM_INQ TCP_INQ
+#endif
+
+#ifdef GRPC_HAVE_MSG_NOSIGNAL
+#define SENDMSG_FLAGS MSG_NOSIGNAL
+#else
+#define SENDMSG_FLAGS 0
+#endif
+
+// TCP zero copy sendmsg flag.
+// NB: We define this here as a fallback in case we're using an older set of
+// library headers that has not defined MSG_ZEROCOPY. Since this constant is
+// part of the kernel, we are guaranteed it will never change/disagree so
+// defining it here is safe.
+#ifndef MSG_ZEROCOPY
+#define MSG_ZEROCOPY 0x4000000
+#endif
+
+#define MAX_READ_IOVEC 64
+
+namespace grpc_event_engine {
+namespace posix_engine {
+
+namespace {
+
+using ::grpc_event_engine::experimental::EndpointConfig;
+using ::grpc_event_engine::experimental::EventEngine;
+using ::grpc_event_engine::experimental::Slice;
+using ::grpc_event_engine::experimental::SliceBuffer;
+
+// A wrapper around sendmsg. It sends \a msg over \a fd and returns the number
+// of bytes sent.
+ssize_t TcpSend(int fd, const struct msghdr* msg, int* saved_errno,
+                int additional_flags = 0) {
+  ssize_t sent_length;
+  do {
+    sent_length = sendmsg(fd, msg, SENDMSG_FLAGS | additional_flags);
+  } while (sent_length < 0 && (*saved_errno = errno) == EINTR);
+  return sent_length;
+}
+
+#ifdef GRPC_LINUX_ERRQUEUE
+
+#define GRPC_LOG_EVERY_N_SEC(n, format, ...)                                  \
+  do {                                                                        \
+    static std::atomic<gpr_cycle_counter> prev{0};                            \
+    gpr_cycle_counter now = gpr_get_cycle_counter();                          \
+    if ((grpc_core::Timestamp::FromCycleCounterRoundDown(now) -               \
+         grpc_core::Timestamp::FromCycleCounterRoundDown(prev.exchange(now))) \
+            .millis() > (n)*1000) {                                           \
+      gpr_log(GPR_INFO, format, __VA_ARGS__);                                 \
+    }                                                                         \
+  } while (0)
+
+#define CAP_IS_SUPPORTED(cap) (prctl(PR_CAPBSET_READ, (cap), 0) > 0)
+
+// Ulimit hard memlock controls per socket limit for maximum locked memory in
+// RAM.
+uint64_t GetUlimitHardMemLock() {
+  static uint64_t kUlimitHardMemLock = []() -> uint64_t {
+    if (CAP_IS_SUPPORTED(CAP_SYS_RESOURCE)) {
+      // hard memlock ulimit is ignored for privileged user.
+      return UINT64_MAX;
+    }
+    // Parses /etc/security/limits.conf file for a line of the following format:
+    // * hard memlock <value>
+    // It extracts <value> and returns it. A value of UINT64_MAX represents
+    // unlimited or infinity. Hard memlock value should be set to allow zerocopy
+    // sendmsgs to succeed. It controls the maximum amount of memory that can be
+    // locked by a socket in RAM.
+    std::ifstream limits_file("/etc/security/limits.conf");
```

Comment:
Could the hard memlock setting be overriden by configurations in `/etc/security/limits.d/`? And are there any supported operating systems that would have this file in another location?

Reply:
Hmm. Interesting. Looks like /etc/security/limits.d/ can have user/program specific overrides and /etc/security/limits.conf is the system default one. I updated the logic to parse every file in /etc/security/limits.d/ + /etc/security/limits.conf to see if any of them have hard memlock set. If thats true, then tx zero copy is allowed to proceed. However its still possible that the memlock may not be set for the program/user running the grpc process. In that case tx zero copy will still see ENOBUFS and we will print a log. I think that should be sufficient.

Reply:
Only linux supports this feature and its specified in these locations /etc/security/limits.d/ or /etc/security/limits.conf